ml-agents/mlagents/trainers/components/reward_signals/gail/signal.py

```diff
@@ -0,0 +1,277 @@
+
+        self.model = GAILModel(
+            policy.model, 128, learning_rate, encoding_size, use_actions
+        )
+        _, self.demonstration_buffer = demo_to_buffer(demo_path, policy.sequence_length)
+        self.has_updated = False
+
+    def evaluate(
+        self, current_info: BrainInfo, next_info: BrainInfo
+    ) -> RewardSignalResult:
+        if len(current_info.agents) == 0:
+            return []
+
+        feed_dict = {
+            self.policy.model.batch_size: len(next_info.vector_observations),
+            self.policy.model.sequence_length: 1,
+            self.model.use_noise: [0],
+        }
+        feed_dict = self.policy.fill_eval_dict(feed_dict, brain_info=current_info)
+        feed_dict[self.model.done_policy] = np.reshape(next_info.local_done, [-1, 1])
+        if self.policy.use_continuous_act:
+            feed_dict[
+                self.policy.model.selected_actions
+            ] = next_info.previous_vector_actions
+        else:
+            feed_dict[
+                self.policy.model.action_holder
+            ] = next_info.previous_vector_actions
+        if self.policy.use_recurrent:
+            if current_info.memories.shape[1] == 0:
+                current_info.memories = self.policy.make_empty_memory(
+                    len(current_info.agents)
+                )
+            feed_dict[self.policy.model.memory_in] = current_info.memories
+        unscaled_reward = self.policy.sess.run(
+            self.model.intrinsic_reward, feed_dict=feed_dict
+        )
+        scaled_reward = unscaled_reward * float(self.has_updated) * self.strength
+        return RewardSignalResult(scaled_reward, unscaled_reward)
+
+    @classmethod
+    def check_config(
+        cls, config_dict: Dict[str, Any], param_keys: List[str] = None
+    ) -> None:
+        """
+        Checks the config and throw an exception if a hyperparameter is missing. GAIL requires strength and gamma 
+        at minimum. 
+        """
+        param_keys = ["strength", "gamma", "demo_path"]
+        super().check_config(config_dict, param_keys)
+
+    def update(self, update_buffer: Buffer, n_sequences: int) -> Dict[str, float]:
+        """
+        Updates model using buffer.
+        :param update_buffer: The policy buffer containing the trajectories for the current policy.
+        :param n_sequences: The total (demo + environment) number of sequences used in each mini batch.
+        :return: The loss of the update.
+        """
+        batch_losses = []
+        n_sequences = max(n_sequences // 2, 1)  # Divide by 2 since we have two buffers
+        possible_demo_batches = (
+            len(self.demonstration_buffer.update_buffer["actions"]) // n_sequences
+        )
+        possible_policy_batches = len(update_buffer["actions"]) // n_sequences
+        possible_batches = min(possible_policy_batches, possible_demo_batches)
+
+        # for reporting
+        kl_loss = []
+        pos = []
```

Comment:
Maybe we want to give these variables more semantically interpretable names?

Reply:
I updated these. TBH, we might be able to remove this since they're only there for printing during debug.

Reply:
In that case I think it makes sense to remove them.